tracing: Add user-definable tracing interface

nslowell opened the pull request:
Create a new tracing option TRACING_USER that allows
the user to define certain user_sys_trace_... functions
to perform whatever work desired for tracing when
tasks are swiched in/out, during isr enter/exit, and when
cpu is idle.

This infrastructure can be useful for plugging into
locally defined tracing tools or any user-specific
debugging environment.

Signed-off-by: Nicholas Lowell <nlowell@lexmark.com>

nslowell committed:
tracing: Add user-definable tracing interface

Create a new tracing option TRACING_USER that allows
the user to define certain user_sys_trace_... functions
to perform whatever work desired for tracing when
tasks are swiched in/out, during isr enter/exit, and when
cpu is idle.

This infrastructure can be useful for plugging into
locally defined tracing tools or any user-specific
debugging environment.

Signed-off-by: Nicholas Lowell <nlowell@lexmark.com>

jukkar commented on subsys/tracing/include/tracing_user.h:
```diff
@@ -0,0 +1,48 @@
+extern "C" {
+#endif
+
+void user_sys_trace_thread_switched_in(struct k_thread *thread);
```
The namespace of these functions look a bit weird, should be place the `user` at the end of the func name, so
`sys_trace_thread_switched_in_user()`?

nslowell replied:
it follows the same naming convention as ctf, so there was precedent for this.

nashif replied:
I agree that user_sys_ is weird, where is this the case for ctf?

nslowell replied:
@nashif 
in ctf_top.c: sys_trace_thread_switched_in() calls ctf_top_thread_switched_in() (where the prefix identifies the subsystem)

I was trying to copy that by having sys_trace_thread_switched_in() call user_sys_trace_thread_switched_in().

Would user_thread_switched_in() be an acceptble name to maintain the prefix precedence?

bbolen replied:
Moved the "user" to the end of the function names.  That seems to be a little more pleasing to the eyes.

nashif commented on doc/guides/tracing/index.rst:
```diff
@@ -140,6 +140,24 @@ and exits (only distinguishes between idle thread, non idle thread and scheduler
+This tracing format allows the user to define functions to perform any work desired
+when a task is switched in or out, when an interrupt is entered or exited, and when the cpu
+is idle.
+
```
Can you please elaborate a bit on the usecase here and why you want to do that? Based on the above this seems like misuse of tracing to do something else not related to tracing, but maybe I am misunderstanding this.

nslowell replied:
Let me know if the below is sufficient.  If so, I can add it:

Examples include
 - simple toggling of GPIO for external scope tracing while minimizing extra cpu load
 - generating/outputting trace data in a non-standard or proprietary format that can not be supported by the other tracing systems

nslowell commented:
I believe the failed test is unrelated and I never had a response from my last comment.

nslowell commented:
@nashif please confirm changes

jukkar reviewed (approved):
LGTM

nashif reviewed (dismissed):
Can you provide a sample/test using this code please?

Commit:
samples: tracing: Add a sample for tracing_user

The user tracing backend allows one to implement weak symbols to perform
user specific tracing, e.g gpio toggling or spi transactions.  This adds
a qemu sample that just does a simple printk for each of the traced
functions, but could be changed to do whatever function the user
desires.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

bbolen commented:
> Can you provide a sample/test using this code please?

added a sample that uses the code which will work in qemu.

bbolen commented:
@nashif ping.  All issues should have been addressed.
